cvmfs/catalog_mgr_impl.h

```diff
@@ -841,6 +841,45 @@ bool AbstractCatalogManager<CatalogT>::IsAttached(const PathString &root_path,
 }
 
 
+template <class CatalogT>
+void AbstractCatalogManager<CatalogT>::StageNestedCatalogAndUnlock(
```

Comment:
i find this function a bit confusing so i tried to rewrite it to understand it better.
maybe you want to use this (just needs a double check that i didnt do any mistake)

```c++
 for (typename NestedCatalogList::const_iterator i = nested_catalogs.begin(),
       iEnd = nested_catalogs.end(); i != iEnd; ++i)
  {
    // only stage the mountpoint that is a correct (partial) match
    // of the requested path (e.g. if path is /a/bc then /a/b is not a match).
    if (path.StartsWith(i->mountpoint) &&
          (path_len == i->mountpoint.GetLength()
          || (path_len > i->mountpoint.GetLength()
              && path.GetChars()[mountpoint_len] == '/'))) {

        // No need to stage, parent with the transaction point is sufficient
        if (!is_listable && (path_len == mountpoint_len)) {
          break;
        }

        Unlock();
        LogCvmfs(kLogCatalog, kLogDebug, "staging nested catalog at %s (%s)",
             i->mountpoint.c_str(), i->hash.ToString().c_str());
        StageNestedCatalogByHash(i->hash, i->mountpoint);
        return;
      }
    }
  Unlock();
}
```